server/src/main/java/org/elasticsearch/transport/TcpTransportKeepAlive.java

```diff
@@ -0,0 +1,204 @@
+                }
+            }
+        });
+    }
+
+    @Override
+    public void close() {
+        lifecycle.moveToStopped();
+        lifecycle.moveToClosed();
+    }
+
+    private class ScheduledPing extends AbstractLifecycleRunnable {
+
+        private final TimeValue pingInterval;
+
+        private final Set<TcpChannel> channels = ConcurrentCollections.newConcurrentSet();
+
+        private final AtomicBoolean isStarted = new AtomicBoolean(false);
+        private volatile long lastPingTime;
+        private ScheduledPing(TimeValue pingInterval) {
+            super(lifecycle, logger);
+            this.pingInterval = pingInterval;
+            this.lastPingTime = System.nanoTime();
+        }
+
+        void ensureStarted() {
+            if (isStarted.get() == false && isStarted.compareAndSet(false, true)) {
+                threadPool.schedule(pingInterval, ThreadPool.Names.GENERIC, this);
+            }
+        }
+
+        void addChannel(TcpChannel channel) {
+            channels.add(channel);
+        }
+
+        void removeChannel(TcpChannel channel) {
+            channels.remove(channel);
+        }
+
+        @Override
+        protected void doRunInLifecycle() {
+            for (TcpChannel channel : channels) {
+                // In the future it is possible that we may want to kill a channel if we have not read from
+                // the channel since the last ping. However, this will need to be backwards compatible with
+                // pre-6.6 nodes that DO NOT respond to pings
+                if (needsKeepAlivePing(channel)) {
+                    sendPing(channel);
+                }
+            }
+            this.lastPingTime = System.nanoTime();
+        }
+
+        @Override
+        protected void onAfterInLifecycle() {
+            try {
+                threadPool.schedule(pingInterval, ThreadPool.Names.GENERIC, this);
+            } catch (EsRejectedExecutionException ex) {
+                if (ex.isExecutorShutdown()) {
+                    logger.debug("couldn't schedule new ping execution, executor is shutting down", ex);
+                } else {
+                    throw ex;
+                }
+            }
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            if (lifecycle.stoppedOrClosed()) {
+                logger.trace("failed to send ping transport message", e);
+            } else {
+                logger.warn("failed to send ping transport message", e);
+            }
+        }
+
+        private boolean needsKeepAlivePing(TcpChannel channel) {
+            TcpChannel.Stats stats = channel.getStats();
+            long writeDelta = stats.lastWriteTime() - lastPingTime;
+            long readDelta = stats.lastReadTime() - lastPingTime;
+            return writeDelta <= 0 || readDelta <= 0;
+        }
+    }
+
+    interface PingSender {
```

Comment:
Just an idea, stuff like this comes up all the time. Can we extract a `AsyncBiFunction` and `AsyncFunction` that is equivalent to `Function` and `BiFunction` then we can cover these cases in the future. In any case we should mark this as a `FunctionalInterface`?

Reply:
`send` is a void method that takes three arguments. It would be `TriConsumer`. Is that something you still want? I did add the `FunctionalInterface`.

Reply:
hmm this is what I had in mind:

```Java
interface AsyncBiFunction<A,B,C> {
  void run(A a, B b, ActionListener<C>);
}
```